README
LIBCDS 2 - A Compressed Data Structure Library
================================================

Rationale
---------
LIBCDS 2 is the re-coding of [LIBCDS][libcds1] with these goals:

 - 64-Bit support to allow for larger, single structures.
 - Speedups
 - Improve code readability
 - Include unit tests ([googletest][googtest])
 - Improve documentation

Further long-term goals include:

 - Introduce missing types to the library.
 - Provide packages for common GNU/Linux distributions, and other OSs.
 - Provide the option of installing it as a shared library (easing compilation and updates).
 - Wrappers for popular programming languages.


Links
-----
Some useful project links are below:

 - [LIBCDS 2 GitHub project][gitproject]
 - [LIBCDS 2 Wiki][wiki]
 - [LIBCDS 2 Issues][issues]
 - [Original LIBCDS Site][libcds1]
 - [LIBCDS 2 Developers Google Group][devgroup]


Guidelines
----------

 - If you make changes to autoconf files (configure.ac or Makefile.am files), please rerun autoconf like so: `autoreconf -vif --no-recursive`.

 - When you commit, it is useful to give a good message.

 - The code tries to follow this style guide [C++ Style Guide][cppstyle].

 - Before commiting, please make sure you run all tests and 'make cpplint/indent' to check for style errors.

[libcds1]: http://libcds.recoded.cl/
[devgroup]: http://groups.google.com/group/libcds-dev/
[googtest]: http://code.google.com/p/googletest/
[gitproject]: https://github.com/fclaude/libcds2
[cppstyle]: http://google-styleguide.googlecode.com/svn/trunk/cppguide.xml
[wiki]: https://github.com/fclaude/libcds2/wiki
[issues]: https://github.com/fclaude/libcds2/issues


ChangeLog
2013-01-10  Francisco Claude  <fclaude@scspc199.cs.uwaterloo.ca>

	fixed makefile to run with the new version of swig also the includes can be installed with shared_lib_install

2013-01-10  Francisco Claude  <fc@akori.cl>

	Adding test for memory manager

	Adding missing tree.cpp file and a small memory manager (work in progress)

	New compilation for tests Updated constants

2012-10-22  Francisco Claude  <fclaude@recoded.cl>

	Adding spire 2012 tutorial

2012-10-20  Francisco Claude  <fclaude@recoded.cl>

	Fixed a memory leak in the test for sequences

	Fixed two leaks in the wavelet tree without pointers

	Load and Save tested for trees and sequences. Valgrind is not showing any problems in Mac OS X, will try it in Linux too.

	Added Save/Load functions for wavelet trees without pointers.

	Adding wavelet trees without pointers. Basic version, needs work on non-standard queries for sequences.

2012-10-18  Francisco Claude  <fclaude@recoded.cl>

	Adding louds and fixed some issues with memory management of permutations

	Adding permutations (Base class and MRRR)

	Merge branch 'master' of github.com:fclaude/libcds2

	Makefile

2012-09-27  Roberto Konow  <rkonowkr@gmail.com>

	Wavelet Tree Basics and more cpplint fixes

	FIX: cpplint of bitsequence,basics,array ADD: WavletTreeNode and WaveletTreeLeaf classes inside WaveletTree Class.

2012-09-25  Roberto Konow  <rkonowkr@gmail.com>

	added WaveletTree definitions and git testing

	testing git header2

	Merge branch 'master' of github.com:fclaude/libcds2

	testing git header...3

2012-09-25  Roberto Konow  <rkonow@ratonera-laptop.(none)>

	testing git header...

2012-09-20  Francisco Claude  <fclaude@recoded.cl>

	Fixed a memory leak

2012-09-19  Francisco Claude  <fclaude@recoded.cl>

	Fixed an array out of bounds in Select/SelectNext

	Fixed makefile and removed a message from array.

2012-09-19  Roberto Konow  <rkonow@ratonera-laptop.(none)>

	temporary makefile

	adding dummy file to keep an empty directory

	Small change to test gccgo

	patada!

2012-09-19  Francisco Claude  <fclaude@recoded.cl>

	Added sample code for Go Wrote a small test to compare C++ vs Go speed Updated .gitignore

	Array out-of-bouds fix and interface for swig improved

	Formatted code

2012-09-19  Roberto Konow  <rkonow@ratonera-laptop.(none)>

	Added Coder Interface and Dummy implementation Fixed Makefile

	forgot to include test

	Added Mapper interface and MapperNone as a dummy implementation

	Fixed: - astyle - memory leak in bitsequenceonelevelrank

2012-09-18  Francisco Claude  <fclaude@scspc199.(none)>

	Added destructor

2012-06-12  Francisco Claude  <fclaude@recoded.cl>

	Fixed a similar bug in Array::BinarySearch (boundary issues)

	fixed a small bug in Array::BinarySearch

2012-06-03  Francisco Claude  <fclaude@recoded.cl>

	Fixed a bug in SelectNext1.

	Fixed a bug in Select1 and SelectNext1.

	Adding simple script for nightly tests runs.

	Removing unused stuff from Makefile

	Fixed bug in optimized version of Access+Rank.

2012-06-02  Francisco Claude  <fclaude@recoded.cl>

	Messed up with the test for bitsequence, now it's fixed.

2012-06-01  Francisco Claude  <fclaude@recoded.cl>

	Fixed formating and style.

	Added Sequence and cleaned up tests.
	- Added the base class for representing sequences
	- Added the restricted select methods to the base BitSequence representation
	- Added (commented) the methods missing in BitSequenceOneLevelRank
	- Cleaned tests for BitSequence, no we only have one test for the base class
	- Added the tests for the class Sequence

2012-05-31  Francisco Claude  <fclaude@recoded.cl>

	Updating README again, removed stuff we are not using.

	Fixed formatting and style. Passing all tests.

	Updated gitignore, and added a script to run all tests.

	Removing reviews page, they are not working.

	Changes: - Fixed some issues with Save/Load - Implemented optimized versions for rank+access and access+rank (BitSequenceOneLevelRank) - Added corresponding tests

2012-05-30  Francisco Claude  <fclaude@recoded.cl>

	Changed the Load and Save functions to use istream and ostream instead of the if/ofstream objects. Also added a Load/Save function for Arrays (new implementation) and BitSequenceOneLevelRank.

	Formatted and revised style.

	Updating README file to what we are actually doing.

	Added: - SelectNext1 to BitSequenceRankOneLevel - Added LowerBoundExp to the array representation (exponential search) - Changed the ceiling function in libcds.h so it avoids using %

2012-05-29  Francisco Claude  <fclaude@recoded.cl>

	Added Select0 for BitSequenceOneLevelRank

	- Added test for BitSequenceOneLevelRank - Finished implementation for Rank1 for BitSequenceOneLevelRank - Finished implementation for Select1 for BitSequenceOneLevelRank - Corrected method names for time.h - Fixed minor things in array, and it's tests - New test fro time.h - All tests are passing at this point

2012-05-28  Francisco Claude  <fclaude@recoded.cl>

	changed a comment

	inlined some functions in array.cpp

2012-05-24  Francisco Claude  <fclaude@recoded.cl>

	adding documentation.

	fixing a lint warning.

	new arrays that optimize access for certain bit-sizes.

2012-03-30  Rodrigo Cánovas  <yigorc@gmail.com>

	GetSize() method ready and only the last part of select1 is missing

	Update src/basic/time.h

	Update time.h and add method to compute CPU usage time

2012-03-29  Rodrigo Cánovas  <yigorc@gmail.com>

	Update src/basic/time.h

2012-03-28  Rodrigo Cánovas  <yigorc@gmail.com>

	Modify Access function, fix Rank (still not tested), and start implement Select 1 using binary search

2012-03-20  Francisco Claude  <fclaude@recoded.cl>

	Adding the one level rank, this is not ready, it may not even compile, but I'm uploading it so Rodrigo Canovas can also work on these files.

2012-02-19  Francisco Claude  <fclaude@recoded.cl>

	Added another test, corrected the previous one.

	Created the first tests for BitSequence. Fixed some issues found by the tests.


src/basic/array.cpp
/********************************************************************************
Copyright (c) 2012, Francisco Claude.
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation
      and/or other materials provided with the distribution.

    * Neither the name of libcds nor the names of its
      contributors may be used to endorse or promote products derived from this
      software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

********************************************************************************/


#include <libcds2/array.h>
#include <libcds2/libcds.h>

#include <fstream>
#include <algorithm>

namespace cds {
namespace basic {

using std::istream;
using std::ostream;

using cds::basic::Array;
using cds::basic::ArrayTpl;

Array *Array::Create(cds_word *A, cds_word i, cds_word j, cds_word bpe) {
  cds_word max_value = 0;
  for (cds_word k = i; k <= j; k++) {
    max_value = max(max_value, A[k]);
  }
  Array *ret = Create(j - i + 1, max(msb(max_value), bpe));
  for (cds_word k = i; k <= j; k++) {
    assert(A[k] <= max_value);
    ret->SetField(k - i, A[k]);
  }
  return ret;
}

Array *Array::Create(const vector<cds_word> &values) {
  cds_word max_value = 0;
  for (cds_word k = 0; k < values.size(); k++) {
    max_value = max(max_value, values[k]);
  }
  Array *ret = Create(values.size(), msb(max_value));
  for (cds_word k = 0; k < values.size(); k++) {
    assert(values[k] <= max_value);
    ret->SetField(k, values[k]);
  }
  return ret;
}

template <cds_word bpe> void ArrayTpl<bpe>::Save(ostream &out) const {
  SaveValue(out, bpe);
  SaveValue(out, length_);
  cds_word uint_length = WordsLength(length_, bpe);
  SaveValue(out, data_, uint_length);
}

template <cds_word bpe> ArrayTpl<bpe>::ArrayTpl(istream &input) {
  length_ = LoadValue<cds_word>(input);
  cds_word uint_length = WordsLength(length_, bpe);
  data_ = LoadValue<cds_word>(input, uint_length);
}

template <cds_word bpe> ArrayTpl<bpe> *ArrayTpl<bpe>::Load(istream &input) {
  cds_word id = LoadValue<cds_word>(input);
  if (id != bpe) {
    return NULL;
  }
  return new ArrayTpl<bpe>(input);
}

template <cds_word bpe> ArrayTpl<bpe>::ArrayTpl(cds_word n) {
  length_ = n;
  InitData();
}

template <cds_word bpe> ArrayTpl<bpe>::ArrayTpl(cds_word *A, cds_word i, cds_word j) {
  length_ = j - i + 1;
  InitData();
  for (cds_word k = i; k <= j; k++) {
    SetField(k - i, A[k]);
  }
}

template <cds_word bpe> cds_word ArrayTpl<bpe>::GetSize() const {
  return sizeof(cds_word) * WordsLength(length_, bpe)
         + sizeof(*this);
}

template <cds_word bpe> cds_word ArrayTpl<bpe>::GetLength() const {
  return length_;
}


template <cds_word bpe> ArrayTpl<bpe>::~ArrayTpl() {
  delete [] data_;
}

template <cds_word bpe> cds_word ArrayTpl<bpe>::GetMax() const {
  cds_word max_value = 0;  // default max
  for (cds_word i = 0; i < length_; i++) {
    max_value = max(max_value, GetField(i));
  }
  return max_value;
}

// Implementation based on the STL lower_bound implementation.
template <cds_word bpe> cds_word ArrayTpl<bpe>::LowerBound(cds_word value, cds_word ini, cds_word fin) const {
  assert(ini <= fin);
  assert(fin <= length_);
  cds_word count, step;
  count = fin - ini;

  while (count > 0) {
    step = count / 2;
    cds_word pos = ini + step;
    if (GetField(pos) < value) {
      ini = pos + 1;
      count -= step + 1;
    } else {
      count = step;
    }
  }
  return ini;
}

template <cds_word bpe> cds_word ArrayTpl<bpe>::LowerBound(cds_word value) const {
  return LowerBound(value, 0, length_);
}

template <cds_word bpe> cds_word ArrayTpl<bpe>::LowerBoundExp(cds_word value, cds_word ini, cds_word fin) const {
  cds_word sum = 1;
  while (ini + sum < fin) {
    if (GetField(ini + sum) >= value) {
      return LowerBound(value, ini, ini + sum);
    }
    ini += sum;
    sum *= 2;
  }
  return LowerBound(value, ini, fin);
}

template <cds_word bpe> cds_word ArrayTpl<bpe>::LowerBoundExp(cds_word value) const {
  return LowerBoundExp(value, 0, length_);
}

// Also based on the STL version.
template <cds_word bpe> cds_word ArrayTpl<bpe>::UpperBound(cds_word value, cds_word ini, cds_word fin) const {
  assert(ini <= fin);
  assert(fin <= length_);
  cds_word count, step;
  count = fin - ini;

  while (count > 0) {
    step = count / 2;
    cds_word pos = ini + step;
    if (!(value < GetField(pos))) {
      ini = pos + 1;
      count -= step + 1;
    } else {
      count = step;
    }
  }
  return ini;
}

template <cds_word bpe> cds_word ArrayTpl<bpe>::UpperBound(cds_word value) const {
  return UpperBound(value, 0, length_);
}

template <cds_word bpe> cds_word ArrayTpl<bpe>::BinarySearch(cds_word value, cds_word ini, cds_word fin) const {
  assert(ini <= fin);
  assert(fin <= length_);
  cds_word pos = LowerBound(value, ini, fin);
  if (pos != fin && !(value < GetField(pos))) {
    return pos;
  } else {
    return fin;
  }
}

template <cds_word bpe> cds_word ArrayTpl<bpe>::BinarySearch(cds_word value) const {
  return BinarySearch(value, 0, length_);
}

template <cds_word bpe> void ArrayTpl<bpe>::InitData() {
  cds_word uint_length = WordsLength(length_, bpe);
  data_ = new cds_word[uint_length];
  for (cds_word i = 0; i < uint_length; i++) {
    data_[i] = 0;
  }
}

template <> cds_word ArrayTpl<32>::GetField(const cds_word position) const {
  assert(position < length_);
  return reinterpret_cast<uint32_t *>(data_)[position];
}

template <> cds_word ArrayTpl<16>::GetField(const cds_word position) const {
  assert(position < length_);
  return reinterpret_cast<uint16_t *>(data_)[position];
}

template <> cds_word ArrayTpl<8>::GetField(const cds_word position) const {
  assert(position < length_);
  return reinterpret_cast<uint8_t *>(data_)[position];
}

template <> cds_word ArrayTpl<4>::GetField(const cds_word position) const {
  assert(position < length_);
  return cds::basic::GetField4(data_, position);
}

template <> cds_word ArrayTpl<2>::GetField(const cds_word position) const {
  assert(position < length_);
  return cds::basic::GetField2(data_, position);
}

template <> cds_word ArrayTpl<1>::GetField(const cds_word position) const {
  assert(position < length_);
  if (cds::basic::BitGet(data_, position)) {
    return 1;
  }
  return 0;
}

template <cds_word bpe> cds_word ArrayTpl<bpe>::GetField(const cds_word position) const {
  assert(position < length_);
  return cds::basic::GetField(data_, bpe, position);
}

template <cds_word bpe> cds_word ArrayTpl<bpe>::SetField(const cds_word position, const cds_word v) {
  assert(position < length_);
  cds::basic::SetField(data_, bpe, position, v);
  return v;
}

Array *Array::Load(istream &input) {
  cds_word id = LoadValue<cds_word>(input);
  size_t pos = input.tellg();
  input.seekg(pos - sizeof(cds_word));
  switch (id) {
    case 0:
      return ArrayTpl<0>::Load(input);
    case 1:
      return ArrayTpl<1>::Load(input);
    case 2:
      return ArrayTpl<2>::Load(input);
    case 3:
      return ArrayTpl<3>::Load(input);
    case 4:
      return ArrayTpl<4>::Load(input);
    case 5:
      return ArrayTpl<5>::Load(input);
    case 6:
      return ArrayTpl<6>::Load(input);
    case 7:
      return ArrayTpl<7>::Load(input);
    case 8:
      return ArrayTpl<8>::Load(input);
    case 9:
      return ArrayTpl<9>::Load(input);
    case 10:
      return ArrayTpl<10>::Load(input);
    case 11:
      return ArrayTpl<11>::Load(input);
    case 12:
      return ArrayTpl<12>::Load(input);
    case 13:
      return ArrayTpl<13>::Load(input);
    case 14:
      return ArrayTpl<14>::Load(input);
    case 15:
      return ArrayTpl<15>::Load(input);
    case 16:
      return ArrayTpl<16>::Load(input);
    case 17:
      return ArrayTpl<17>::Load(input);
    case 18:
      return ArrayTpl<18>::Load(input);
    case 19:
      return ArrayTpl<19>::Load(input);
    case 20:
      return ArrayTpl<20>::Load(input);
    case 21:
      return ArrayTpl<21>::Load(input);
    case 22:
      return ArrayTpl<22>::Load(input);
    case 23:
      return ArrayTpl<23>::Load(input);
    case 24:
      return ArrayTpl<24>::Load(input);
    case 25:
      return ArrayTpl<25>::Load(input);
    case 26:
      return ArrayTpl<26>::Load(input);
    case 27:
      return ArrayTpl<27>::Load(input);
    case 28:
      return ArrayTpl<28>::Load(input);
    case 29:
      return ArrayTpl<29>::Load(input);
    case 30:
      return ArrayTpl<30>::Load(input);
    case 31:
      return ArrayTpl<31>::Load(input);
    case 32:
      return ArrayTpl<32>::Load(input);
    case 33:
      return ArrayTpl<33>::Load(input);
    case 34:
      return ArrayTpl<34>::Load(input);
    case 35:
      return ArrayTpl<35>::Load(input);
    case 36:
      return ArrayTpl<36>::Load(input);
    case 37:
      return ArrayTpl<37>::Load(input);
    case 38:
      return ArrayTpl<38>::Load(input);
    case 39:
      return ArrayTpl<39>::Load(input);
    case 40:
      return ArrayTpl<40>::Load(input);
    case 41:
      return ArrayTpl<41>::Load(input);
    case 42:
      return ArrayTpl<42>::Load(input);
    case 43:
      return ArrayTpl<43>::Load(input);
    case 44:
      return ArrayTpl<44>::Load(input);
    case 45:
      return ArrayTpl<45>::Load(input);
    case 46:
      return ArrayTpl<46>::Load(input);
    case 47:
      return ArrayTpl<47>::Load(input);
    case 48:
      return ArrayTpl<48>::Load(input);
    case 49:
      return ArrayTpl<49>::Load(input);
    case 50:
      return ArrayTpl<50>::Load(input);
    case 51:
      return ArrayTpl<51>::Load(input);
    case 52:
      return ArrayTpl<52>::Load(input);
    case 53:
      return ArrayTpl<53>::Load(input);
    case 54:
      return ArrayTpl<54>::Load(input);
    case 55:
      return ArrayTpl<55>::Load(input);
    case 56:
      return ArrayTpl<56>::Load(input);
    case 57:
      return ArrayTpl<57>::Load(input);
  }
  return NULL;
}

Array *Array::Create(cds_word n, cds_word bpe) {
  switch (bpe) {
    case 0:
      return new ArrayTpl<0>(n);
    case 1:
      return new ArrayTpl<1>(n);
    case 2:
      return new ArrayTpl<2>(n);
    case 3:
      return new ArrayTpl<3>(n);
    case 4:
      return new ArrayTpl<4>(n);
    case 5:
      return new ArrayTpl<5>(n);
    case 6:
      return new ArrayTpl<6>(n);
    case 7:
      return new ArrayTpl<7>(n);
    case 8:
      return new ArrayTpl<8>(n);
    case 9:
      return new ArrayTpl<9>(n);
    case 10:
      return new ArrayTpl<10>(n);
    case 11:
      return new ArrayTpl<11>(n);
    case 12:
      return new ArrayTpl<12>(n);
    case 13:
      return new ArrayTpl<13>(n);
    case 14:
      return new ArrayTpl<14>(n);
    case 15:
      return new ArrayTpl<15>(n);
    case 16:
      return new ArrayTpl<16>(n);
    case 17:
      return new ArrayTpl<17>(n);
    case 18:
      return new ArrayTpl<18>(n);
    case 19:
      return new ArrayTpl<19>(n);
    case 20:
      return new ArrayTpl<20>(n);
    case 21:
      return new ArrayTpl<21>(n);
    case 22:
      return new ArrayTpl<22>(n);
    case 23:
      return new ArrayTpl<23>(n);
    case 24:
      return new ArrayTpl<24>(n);
    case 25:
      return new ArrayTpl<25>(n);
    case 26:
      return new ArrayTpl<26>(n);
    case 27:
      return new ArrayTpl<27>(n);
    case 28:
      return new ArrayTpl<28>(n);
    case 29:
      return new ArrayTpl<29>(n);
    case 30:
      return new ArrayTpl<30>(n);
    case 31:
      return new ArrayTpl<31>(n);
    case 32:
      return new ArrayTpl<32>(n);
    case 33:
      return new ArrayTpl<33>(n);
    case 34:
      return new ArrayTpl<34>(n);
    case 35:
      return new ArrayTpl<35>(n);
    case 36:
      return new ArrayTpl<36>(n);
    case 37:
      return new ArrayTpl<37>(n);
    case 38:
      return new ArrayTpl<38>(n);
    case 39:
      return new ArrayTpl<39>(n);
    case 40:
      return new ArrayTpl<40>(n);
    case 41:
      return new ArrayTpl<41>(n);
    case 42:
      return new ArrayTpl<42>(n);
    case 43:
      return new ArrayTpl<43>(n);
    case 44:
      return new ArrayTpl<44>(n);
    case 45:
      return new ArrayTpl<45>(n);
    case 46:
      return new ArrayTpl<46>(n);
    case 47:
      return new ArrayTpl<47>(n);
    case 48:
      return new ArrayTpl<48>(n);
    case 49:
      return new ArrayTpl<49>(n);
    case 50:
      return new ArrayTpl<50>(n);
    case 51:
      return new ArrayTpl<51>(n);
    case 52:
      return new ArrayTpl<52>(n);
    case 53:
      return new ArrayTpl<53>(n);
    case 54:
      return new ArrayTpl<54>(n);
    case 55:
      return new ArrayTpl<55>(n);
    case 56:
      return new ArrayTpl<56>(n);
    case 57:
      return new ArrayTpl<57>(n);
    case 58:
      return new ArrayTpl<58>(n);
  }
  return NULL;
}
};
};
